api/v1alpha4/clusterclass_webhooks.go

```diff
@@ -0,0 +1,154 @@
+		}
+		if len(in.Spec.Worker.Deployments[i].Template.Infrastructure.Ref.Namespace) == 0 {
+			in.Spec.Worker.Deployments[i].Template.Infrastructure.Ref.Namespace = in.Namespace
+		}
+	}
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateCreate() error {
+	return in.validate(nil)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateUpdate(old runtime.Object) error {
+	oldClusterClass, ok := old.(*ClusterClass)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterClass but got a %T", old))
+	}
+	return in.validate(oldClusterClass)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateDelete() error {
+	return nil
+}
+
+func (in *ClusterClass) validate(old *ClusterClass) error {
+	var allErrs field.ErrorList
+
+	// ensure all the references are within the same namespace
+	if in.Spec.Infrastructure.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs, field.Invalid(field.NewPath("spec", "infrastructure", "ref", "namespace"), in.Spec.Infrastructure.Ref.Namespace, "must match metadata.namespace"),
+		)
+	}
+	if in.Spec.ControlPlane.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs, field.Invalid(field.NewPath("spec", "controlPlane", "ref", "namespace"), in.Spec.ControlPlane.Ref.Namespace, "must match metadata.namespace"),
+		)
+	}
+	for _, class := range in.Spec.Worker.Deployments {
+		if class.Template.Bootstrap.Ref.Namespace != in.Namespace {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments", "template", "bootstrap", "ref", "namespace"), class.Template.Bootstrap.Ref.Namespace, "must match metadata.namespace"),
+			)
+		}
+		if class.Template.Infrastructure.Ref.Namespace != in.Namespace {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments", "infrastructure", "bootstrap", "ref", "namespace"), class.Template.Infrastructure.Ref.Namespace, "must match metadata.namespace"),
+			)
+		}
+	}
+
+	// Ensure MachineDeployment class are unique.
+	classNames := make(map[string]struct{})
+	for _, class := range in.Spec.Worker.Deployments {
+		if _, ok := classNames[class.Class]; ok {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments"), class, fmt.Sprintf("MachineDeployment class %q is defined more than once.", class.Class)),
+			)
+		}
+		classNames[class.Class] = struct{}{}
+	}
+
+	// In case of updates
+	if old != nil {
+		// Makes sure all the old MachineDeployment class are still there (only MachineDeployment class addition are allowed).
+		oldClassNames := make(map[string]struct{})
+		for _, oldClass := range old.Spec.Worker.Deployments {
```

Comment:
Should we add an utility function on the struct that returns a strings.Set of all the unique names?

Reply:
It seems it's used a lot, which could help reduce the complexity of this code

Reply:
If we get a set of names, we won't be able to detect duplicates, which is another use case where we are doing this.
Also, I see those few of those mutability constraint to go away in the near future, so I don't think we should create utility functions.

Reply:
If we get a set and the length of the list is greater than the set, we have dupes

Reply:
I think with the set we're not able to return a precise error message which mentions the duplicate classes (or at least we then have to add some new logic for this) 
(also referring to the other use case above)

Reply:
I'll defer this code optimisation (or better the removal of some of those constraints) to later, this PR is already getting bigger than expected...